src/rgw/rgw_data_sync.cc

```diff
@@ -1749,6 +1758,32 @@ class RGWDataSyncShardCR : public RGWCoroutine {
             marker_tracker->try_update_high_marker(log_iter->log_id, 0, log_iter->log_timestamp);
             continue;
           }
+          if (!log_iter->entry.gen) {
+            yield {
+              rgw_bucket_index_marker_info remote_info;
+              BucketIndexShardsManager remote_markers;
+              retcode = rgw_read_remote_bilog_info(sync_env->dpp, sc->conn, source_bs.bucket,
+                            remote_info, remote_markers, null_yield);
+
+              if (retcode < 0) {
+                tn->log(1, SSTR(" rgw_read_remote_bilog_info failed with retcode=" << retcode));
+                return retcode;
+              }
+              for (const auto& each : remote_info.gen_numshards) {
+                for (int sid = 0; sid < each.second; sid++) {
+                  rgw_bucket_shard bs(source_bs.bucket, sid);
+                  error_repo = datalog_oid_for_error_repo(bs);
+                  tn->log(10, SSTR("writing shard_id " << sid << "of gen" << each.first << " to error repo for retry"));
+                  call(rgw::error_repo::write_cr(sync_env->store->svc()->rados, error_repo,
+                                                      rgw::error_repo::encode_key(bs, each.first),
+                                                      ceph::real_time{}));
+                  if (retcode < 0) {
+                    tn->log(0, SSTR("ERROR: failed to log sync failure in error repo: retcode=" << retcode));
+                  }
+                }
+              }
+            }
+          }
```

Comment:
i think it's worth factoring this whole block out to a new coroutine class

i also don't think this logic belongs in the for loop over `log_entries`. we don't expect to see these full sync obligations from another zone's datalog, but from the error repo for things that failed during data full sync. so i think this belongs in the loop over `error_entries` above

and once we've successfully written each shard entry to the error repo, we'll want to remove the original 'full' obligation from the error repo with `rgw::error_repo::remove_cr()`